src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1388,7 +1372,20 @@ void Cancel(bool throwExceptions)
             if (connector == null)
                 return;
 
-            connector.CancelRequest(throwExceptions);
+            try
+            {
+                if (connector.CancelRequest(throwExceptions: true))
+                {
+                    if (connector.Settings.CancellationTimeout > 0)
```

Comment:
This logic can maybe move inside NpgsqlConnector.CancelRequest. This would deduplicate the logic in NpgsqlDataReader.Cancel, will probably simplify things (no need to communicate out whether cancellation actually was performed), would obviate needing to manually trigger the WriteCts when the timeout occurs, etc.